### DISCUSSION TITLE
Properly handle broken resource

### DISCUSSION THREAD

**Discussion by PedroMPagani** [xetorthio/jedis#3182] [Q&A] [answered]
Currently from what I see one of the best ways to handle broken resources is mainly the testOnBorrow, which basically would execute a ping before returning the resource, but that can be a + on the latency of the call, even tho small, could be avoided?

I'm mainly trying to find the best ways to make sure that the operations would still execute even if it encounters a broken Jedis connection, so far came up with something like:

```java
public void executeProcess(Consumer<Jedis> consumer){
    Jedis jedis = this.pool.getResource();
    try {
       consumer.accept(jedis);
    } finally {
      this.pool.returnResource(jedis);
      return;
    } catch(Exception exception){
      this.pool.returnBrokenResource(jedis);
      executeProcess(consumer);
      return;
    }
  }
```

Also including a pubsub safety in case the redis server is restarted, all pubsubs would basicaly disconnect, and if the software does not handle it properly it would basically break, does this seems tangeable or something good?

```java
Thread thread = new Thread(()->{
while(true){
 // register pubsub here, if the subscription throws an exception, try to subscribe again?
}
});
thread.setDaemon(true);
thread.start();
```

Any ideas on hw to improve or how to properly handle the fails and still make sure the op executes?

**Comment by sazzad16** [comment, marked as answer]
For the first part:

Try UnifiedJedis with RetryableCommandExecutor.

```
final int maxAttempts = 5;
final Duration maxRetriesDuration = Duration.ofSeconds(2);

PooledConnectionProvider pooledConnectionProvider = new PooledConnectionProvider(hostAndPort, clientConfig);
RetryableCommandExecutor retryableCommandExecutor = new RetryableCommandExecutor(pooledConnectionProvider, maxAttempts, maxRetriesDuration);
UnifiedJedis jedis = new UnifiedJedis(retryableCommandExecutor);
```

---

For the second part:

Yes, subscribe is not a strength of this library. We welcome any contribution :slightly_smiling_face:

**Reply by PedroMPagani** [reply]
> For the first part:
> 
> Try UnifiedJedis with RetryableCommandExecutor.
> 
> ```
> final int maxAttempts = 5;
> final Duration maxRetriesDuration = Duration.ofSeconds(2);
> 
> PooledConnectionProvider pooledConnectionProvider = new PooledConnectionProvider(hostAndPort, clientConfig);
> RetryableCommandExecutor retryableCommandExecutor = new RetryableCommandExecutor(pooledConnectionProvider, maxAttempts, maxRetriesDuration);
> UnifiedJedis jedis = new UnifiedJedis(retryableCommandExecutor);
> ```
> 
> ---
> 
> For the second part:
> 
> Yes, subscribe is not a strength of this library. We welcome any contribution :slightly_smiling_face: 

Is it still thread safe?

**Reply by sazzad16** [reply]
> Is it still thread safe?

Yes.

**Reply by PedroMPagani** [reply]
> > Is it still thread safe?
> 
> Yes.

What type of class would the clientConfig be? Do you have any example for me?

**Reply by sazzad16** [reply]
> What type of class would the clientConfig be?

This is any implementation of `JedisClientConfig`.

> Do you have any example for me?

If you are going to use all default values, you can even do `new JedisClientConfig() { }`.

There are many uses in out tests. You can search those by `DefaultJedisClientConfig`.

Example:

https://github.com/redis/jedis/blob/870897d86f2d655d330c7066bee3606d3a4f8615/src/test/java/redis/clients/jedis/JedisTest.java#L50-L60

**Reply by PedroMPagani** [reply]
Hey, so I gave it a try, is there any specific reason why the first call would take a long time?

0 took 21.741298

```java
            final int maxAttempts = 5;
            final Duration maxRetriesDuration = Duration.ofSeconds(2);
            ConnectionFactory connectionFactory = new ConnectionFactory(new HostAndPort("10.8.1.1",6379), new BidJedisClientConfig());
            PooledConnectionProvider pooledConnectionProvider = new PooledConnectionProvider(connectionFactory, getConfig());

            RetryableCommandExecutor retryableCommandExecutor = new RetryableCommandExecutor(pooledConnectionProvider, maxAttempts, maxRetriesDuration);
            
            try(UnifiedJedis unifiedJedis = new UnifiedJedis(retryableCommandExecutor)){
                double avg = 0;
                for (int i = 0; i < 100; i++) {
                    Long nano = System.nanoTime();
                    unifiedJedis.publish("test".getBytes(StandardCharsets.UTF_8),"message".getBytes());
                    Long finish = System.nanoTime();
                    avg = (avg + (finish-nano)/1000000.0)/2;
                    System.out.println(i + " took " + (finish-nano)/1000000.0);
                }
                System.out.println("Avg -> " + avg);
            } catch (Exception exception){
                exception.printStackTrace();
            }
```

```java
public class BidJedisClientConfig implements JedisClientConfig {
    
    @Override
    public String getClientName() {
        return "BidController";
    }

    @Override
    public HostAndPortMapper getHostAndPortMapper() {
        return hostAndPort -> new HostAndPort("10.8.1.1",6379);
    }

}
```



Wouldn't the pool open the connection right when its created if my config says min idle?
```java
    public static GenericObjectPoolConfig<Connection> getConfig() {
        GenericObjectPoolConfig<Connection> poolConfig = new GenericObjectPoolConfig<>();
        poolConfig.setMaxTotal(Integer.MAX_VALUE);
        poolConfig.setMaxIdle(128);
        poolConfig.setMinIdle(4);
        poolConfig.setMinEvictableIdleTimeMillis(Duration.ofSeconds(60).toMillis());
        poolConfig.setTimeBetweenEvictionRunsMillis(Duration.ofSeconds(30).toMillis());
        poolConfig.setNumTestsPerEvictionRun(3);
        poolConfig.setBlockWhenExhausted(true);
        return poolConfig;
    }
```

**Reply by sazzad16** [reply]
> is there any specific reason why the first call would take a long time?

It suppose not to.

**Reply by PedroMPagani** [reply]
Anything wrong in my code that would potentially cause it to?

(From what I am seeing, this also happens in JedisPool, if I have some sort of fixed executor the first call on each thread takes a lot more time)

**Reply by sazzad16** [reply]
> this also happens in JedisPool

In that case, can't help you further. JedisPool has been here for a quite long time and considered pretty stable.

<strike>You may check your jvm, cpu network, etc.</strike>
You may check your jvm, cpu, network, etc.

**Reply by PedroMPagani** [reply]
> > this also happens in JedisPool
> 
> In that case, can't help you further. JedisPool has been here for a quite long time and considered pretty stable.
> 
> You may check your jvm, cpu network, etc.

Do you know by any chance what configs in redis side that could cause this? Or this would be 100% because of the JVM, I can tell you I have tcpbacklog at 65k, I have already disabled the transparent pages in the node of the redis, would I have to also disable it in the prod nodes?

**Reply by sazzad16** [reply]
I don't recall any Jedis/Redis specific issue which causes only the first call to be slow. Sorry :slightly_frowning_face:

**Reply by PedroMPagani** [reply]
> I don't recall any Jedis/Redis specific issue which causes only the first call to be slow. Sorry :slightly_frowning_face: 

I understand, what would I be looking for in the JVM to identify what's causing it?

**Reply by sazzad16** [reply]
Random weird stuffs, i.e.
- if the behavior is true for other libraries using sockets, specifically synchronous sockets (java.io)
- try a different version
- try a different implementation
